Certain Factories are not working since upgrade to 5.0.0-M1

Since the upgrade to 5.0.0-M1 certain factories are not working anymore. 

The factories identified and not working:
- https://github.com/bitnami/bitnami-docker-express/tree/che
- https://github.com/bitnami/bitnami-docker-swift/tree/che
- https://github.com/bitnami/bitnami-docker-laravel/tree/che
- https://github.com/bitnami/bitnami-docker-symfony/tree/che
- https://github.com/bitnami/bitnami-docker-java-play/tree/che
- https://github.com/bitnami/bitnami-docker-codeigniter/tree/che
- https://github.com/bitnami/bitnami-docker-rails/tree/che

The factories identified hanging:
- vertx: http://beta.codenvy.com/f?id=9skz1321dxbqpj5j
![screen shot 2016-09-19 at 19 22 18](https://cloud.githubusercontent.com/assets/1636769/18641987/371ebc74-7e9f-11e6-8a22-d128227994cd.png)
- Eclipse Che getting started webpage: https://www.eclipse.org/che/getting-started/cloud/


Factories accepted but landing with an error:
- Gitlab: https://beta.codenvy.com/f/?url=https://gitlab.com/benoitf/simple-project.git
![screen shot 2016-09-19 at 19 25 16](https://cloud.githubusercontent.com/assets/1636769/18641903/e11ac674-7e9e-11e6-9783-b2693e63a8b9.png)

```
2016-09-19 19:04:41,700[spaceManager-98]  [ERROR] [o.e.c.a.w.s.WorkspaceManager 665]    - Start of environment java failed. Error: Error response from docker API, status: 500, message: Error response from daemon: Error downloading remote context http://beta.codenvy.com/api/recipe/recipega4hxep75ipg04yo/script: Got HTTP status code >= 400: 401 Unauthorized

org.eclipse.che.api.core.ServerException: Start of environment java failed. Error: Error response from docker API, status: 500, message: Error response from daemon: Error downloading remote context http://beta.codenvy.com/api/recipe/recipega4hxep75ipg04yo/script: Got HTTP status code >= 400: 401 Unauthorized

	at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.start(WorkspaceRuntimes.java:277) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	at org.eclipse.che.api.workspace.server.WorkspaceManager.lambda$performAsyncStart$1(WorkspaceManager.java:649) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	at org.eclipse.che.api.workspace.server.WorkspaceManager$$Lambda$171/922808511.run(Unknown Source) [che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:28) [che-core-commons-lang-5.0.0-M1.jar:5.0.0-M1]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
	at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
Caused by: org.eclipse.che.api.core.ServerException: Error response from docker API, status: 500, message: Error response from daemon: Error downloading remote context http://beta.codenvy.com/api/recipe/recipega4hxep75ipg04yo/script: Got HTTP status code >= 400: 401 Unauthorized

	at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startEnvironmentQueue(CheEnvironmentEngine.java:738) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	at org.eclipse.che.api.environment.server.CheEnvironmentEngine.start(CheEnvironmentEngine.java:234) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.start(WorkspaceRuntimes.java:246) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	... 6 common frames omitted
Caused by: org.eclipse.che.api.core.ServerException: Error response from docker API, status: 500, message: Error response from daemon: Error downloading remote context http://beta.codenvy.com/api/recipe/recipega4hxep75ipg04yo/script: Got HTTP status code >= 400: 401 Unauthorized

	at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startInstance(CheEnvironmentEngine.java:827) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startEnvironmentQueue(CheEnvironmentEngine.java:673) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	... 8 common frames omitted
Caused by: org.eclipse.che.api.core.ServerException: Error response from docker API, status: 500, message: Error response from daemon: Error downloading remote context http://beta.codenvy.com/api/recipe/recipega4hxep75ipg04yo/script: Got HTTP status code >= 400: 401 Unauthorized

	at org.eclipse.che.plugin.docker.machine.ComposeMachineProviderImpl.startService(ComposeMachineProviderImpl.java:335) ~[che-plugin-docker-machine-5.0.0-M1.jar:5.0.0-M1]
	at org.eclipse.che.api.environment.server.CheEnvironmentEngine.lambda$startEnvironmentQueue$7(CheEnvironmentEngine.java:663) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	at org.eclipse.che.api.environment.server.CheEnvironmentEngine$$Lambda$189/2017672467.startMachine(Unknown Source) ~[na:na]
	at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startInstance(CheEnvironmentEngine.java:775) ~[che-core-api-workspace-5.0.0-M1.jar:5.0.0-M1]
	... 9 common frames omitted
```

--- a/plugins/plugin-urlfactory/pom.xml
+++ b/plugins/plugin-urlfactory/pom.xml
@@ -62,6 +62,10 @@
         </dependency>
         <dependency>
             <groupId>org.eclipse.che.core</groupId>
+            <artifactId>che-core-api-factory</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.eclipse.che.core</groupId>
             <artifactId>che-core-api-factory-shared</artifactId>
         </dependency>
         <dependency>

--- a/plugins/plugin-urlfactory/src/main/java/com/codenvy/plugin/urlfactory/URLFactoryBuilder.java
+++ b/plugins/plugin-urlfactory/src/main/java/com/codenvy/plugin/urlfactory/URLFactoryBuilder.java
@@ -15,7 +15,9 @@
 package com.codenvy.plugin.urlfactory;
 
 import com.google.common.base.Strings;
+import com.google.common.io.CharStreams;
 
+import org.eclipse.che.api.factory.server.FactoryMessageBodyAdapter;
 import org.eclipse.che.api.factory.shared.dto.FactoryDto;
 import org.eclipse.che.api.workspace.shared.dto.EnvironmentDto;
 import org.eclipse.che.api.workspace.shared.dto.EnvironmentRecipeDto;
@@ -26,6 +28,12 @@ import org.eclipse.che.dto.server.DtoFactory;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
 import static java.util.Collections.singletonList;
 import static java.util.Collections.singletonMap;
 import static org.eclipse.che.dto.server.DtoFactory.newDto;
@@ -61,6 +69,9 @@ public class URLFactoryBuilder {
     @Inject
     private URLFetcher URLFetcher;
 
+    @Inject
+    private FactoryMessageBodyAdapter factoryAdapter;
+
     /**
      * Build a default factory using the provided json file or create default one
      *
@@ -74,6 +85,14 @@ public class URLFactoryBuilder {
         if (createFactoryParams != null && createFactoryParams.codenvyJsonFileLocation() != null) {
             String factoryJsonContent = URLFetcher.fetch(createFactoryParams.codenvyJsonFileLocation());
             if (!Strings.isNullOrEmpty(factoryJsonContent)) {
+                // Adapt an old factory format to a new one if necessary
+                try {
+                    final ByteArrayInputStream contentStream = new ByteArrayInputStream(factoryJsonContent.getBytes(UTF_8));
+                    final InputStream newStream = factoryAdapter.adapt(contentStream);
+                    factoryJsonContent = CharStreams.toString(new InputStreamReader(newStream, UTF_8));
+                } catch (IOException x) {
+                    throw new IllegalStateException(x.getLocalizedMessage(), x);
+                }
                 return DtoFactory.getInstance().createDtoFromJson(factoryJsonContent, FactoryDto.class);
             }
         }

--- a/plugins/plugin-urlfactory/src/test/java/com/codenvy/plugin/urlfactory/URLFactoryBuilderTest.java
+++ b/plugins/plugin-urlfactory/src/test/java/com/codenvy/plugin/urlfactory/URLFactoryBuilderTest.java
@@ -14,6 +14,7 @@
  */
 package com.codenvy.plugin.urlfactory;
 
+import org.eclipse.che.api.factory.server.FactoryMessageBodyAdapter;
 import org.eclipse.che.api.factory.shared.dto.FactoryDto;
 import org.eclipse.che.api.workspace.shared.dto.EnvironmentDto;
 import org.eclipse.che.api.workspace.shared.dto.EnvironmentRecipeDto;
@@ -33,6 +34,7 @@ import static java.lang.Boolean.FALSE;
 import static java.util.Collections.singletonList;
 import static java.util.Collections.singletonMap;
 import static org.eclipse.che.dto.server.DtoFactory.newDto;
+import static org.mockito.Matchers.any;
 import static org.mockito.Mockito.when;
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertNull;
@@ -56,6 +58,9 @@ public class URLFactoryBuilderTest {
      */
     @Mock
     private URLFetcher URLFetcher;
+
+    @Mock
+    private FactoryMessageBodyAdapter factoryAdapter;
 
     /**
      * Tested instance.
@@ -151,6 +156,8 @@ public class URLFactoryBuilderTest {
      */
     @Test
     public void checkWithCustomFactoryJsonFile() throws Exception {
+
+        when(factoryAdapter.adapt(any())).thenAnswer(inv -> inv.getArguments()[0]);
 
         WorkspaceConfigDto workspaceConfigDto = newDto(WorkspaceConfigDto.class);
         FactoryDto templateFactory = newDto(FactoryDto.class).withV("4.0").withName("florent").withWorkspace(workspaceConfigDto);

